File: Python solutions/234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        # use extra list
        # time complexity: o(n)
        # space complexity: o(n)
        # cur = head
        # vals = []
        # while cur != None:
        #     vals.append(cur.val)
        #     cur = cur.next
        # front = 0
        # back = len(vals) - 1 
        # while front < back:
        #     if vals[front] != vals[back]: return False
        #     front += 1
        #     back -= 1
        # return True
        
        # use in place comparision by reverse the second half 
        # time complexity: o(n)
        # space complexity: o(1)
        def find_first_half_end(head) -> ListNode:
            fast = slow = head
            while fast.next != None and fast.next.next != None:
                fast = fast.next.next
                slow = slow.next
            return slow
        
        def reverse(head) -> ListNode:
            prev, curr = None, head
            while curr != None:
                curr.next, prev, curr = prev, curr, curr.next
            return prev
        
        if head == None or head.next == None: return True
        
        # find the end of first half and reverse the second half
        firstHalfEnd = secondHalfStart = ListNode()
        firstHalfEnd = find_first_half_end(head)
        secondHalfStart = reverse(firstHalfEnd.next)
        
        # compare first half with the second half
        p1, p2, result = head, secondHalfStart, True
        while result and p2 != None:
            if p1.val != p2.val: result = False
            p1 = p1.next
            p2 = p2.next
        
        # reverse the second half back
        firstHalfEnd.next = reverse(secondHalfStart)
        
        return result

        
        
        
        

    
    

File: Java solutions/234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        /*
            use a extra List 
            time complexity: o(n) 
            space complexity: o(n) for using arraylist
        */ 
        // convert linked list to arraylist
        // List<Integer> vals = new ArrayList<>();
        // ListNode cur = head;
        // while(cur != null){
        //     vals.add(cur.val);
        //     cur = cur.next;
        // }
        // // System.out.println(vals.toString()); // debug print
        // // use two pointer to check for palindrome
        // int front = 0;
        // int back = vals.size() - 1;
        // while (front < back){
        //     if (!vals.get(front).equals(vals.get(back))) return false;
        //     front++;
        //     back--;
        // }
        // return true;
        
        /*  
            in place comparision by reverse the second half
            time complexity: o(4n) -> o(n)
            space complexity: o(1)  
        */
        // find the end of the first half and reverse the second half
        ListNode firstHalfEnd = find_end_of_first_half(head);
        ListNode secondHalfStart = reverse_linked_list(firstHalfEnd.next);
        // compare the first half with the second half
        ListNode p1 = head;
        ListNode p2 = secondHalfStart;
        boolean result = true;
        while (result && p2 != null){
            if (p2.val != p1.val) result = false;
            p1 = p1.next;
            p2 = p2.next;
        }
        // reverse back the second half
        firstHalfEnd.next = reverse_linked_list(secondHalfStart);
        return result;
    }
    
    private ListNode find_end_of_first_half(ListNode head){
        ListNode fast = head;
        ListNode slow = head;
        while(fast.next != null && fast.next.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
    
    private ListNode reverse_linked_list(ListNode head){
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null){
            ListNode temp_next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp_next;
        }
        return prev;
    }
}
